### ISSUE TITLE
Allow group name templating

### ISSUE THREAD

**Issue by flo-everett-xero** [renovatebot/renovate#19011]
### What would you like Renovate to be able to do?

It is common practice in Nuget packages to have "hierarchical" package names - e.g. 
```
Microsoft.Extensions.Configuration.Json
Microsoft.Extensions.Configuration.Abstractions
Microsoft.Extensions.Configuration.Binder
```
which are all members of a `Microsoft.Extensions.Configuration` group, and should all be updated together, or
```
AWSSDK.Core
AWSSDK.SQS
AWSSDK.SecurityToken
AWSSDK.SimpleNotificationService
```
which are all members of a `AWSSDK` group and should be updated together, or many many more examples.

It would be very useful to be able to group packages like this without individually defining every single package group, but rather using regex and templates. I would envision the config would look something like 

```
{
  "matchPackagePatterns": [
     "(?<groupName>Microsoft[.]Extensions[.][^.]+)([.].+)*"
  ]
}
```
or
```
{
  "matchPackagePrefixes": [
     "Microsoft.Extensions."
  ],
  "groupNamePattern": "Microsoft[.]Extensions[.][^.]+"
}
```
these would allow grouping, e.g, all the `Microsoft.Extensions.Configuration` packages, all the `Microsoft.Extensions.Logging` packages, and all the `Microsoft,Extensions.Hosting` packages into separate groups.

### If you have any ideas on how this should be implemented, please tell us here.

as groups are currently implemented only as basic strings, I'm not sure where one would start with this one

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by viceice** [comment]
most of them should already be grouped by sourceUrl. 🤔

do you use a custom nuget registry?

**Comment by flo-everett-xero** [comment]
we do. does that change this behaviour?

**Comment by viceice** [comment]
> we do. does that change this behaviour?

Yes, if your custom nuget registry doesn't support the new v3 api, then renovate can't fetch the nuspec file which contains the repo source url for grouping by url.

**Comment by viceice** [comment]
no, those are the versions endpoints. here is how we fetch the nuspec file for repo url:

https://github.com/renovatebot/renovate/blob/11a96b72a6b2564576edbc50086dd3cb28b4a390/lib/modules/datasource/nuget/v3.ts#L161-L181

`https://api.nuget.org/v3` is the official nuget registry, not a custom. So you are only using that?

**Comment by rarkins** [comment]
@flo-everett-xero in case it wasn't obvious:
- @viceice is suggesting that Renovate could/should perhaps be grouping by the `sourceUrl` properly of a package (same source repo, same group)
- He's wondering if the registry you're using is returning a source repo URL or not

**Comment by flo-everett-xero** [comment]
thank you for your responses - apologies for the late reply, somehow I never received the emails for your comments. 
Apologies for my lack of clarity/confusion - we have two different package sources. 
- `https://api.nuget.org/v3` for any public packages (and all packages that I would expect this feature to work against)
- `https://artifactory.internal-domain.com/api/nuget/nuget` for all internal packages, which I believe is v2 (although some digging has revealed that it also exposes the v3 api, so using that instead would be possible)

It would make sense to me that a new feature would not work against v2 of the api.
that being said, the artifactory v3 `index.json` doesn;'t seem to expose a `PackageBaseAddress` at all, so even on v3 it probably wouldn't work for that registry
